### ISSUE TITLE
Shuldn't an interface go with the protocol ?

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11032]
|<img alt="pupeno's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pupeno reported|
|-|-|
|Trac ID|trac#1489|
|Type|enhancement|
|Created|2006-02-16 01:04:58Z|

Attachments:

* [twistedquotes.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/047/04754785b32160c5787f32875f2dc730c10cbbcb/14987cdc7fb65c51804ea342d1dce82e6b04cfbb.diff) (2471 bytes) - added by exarkun on 2008-06-15 20:22:34Z - move IQuoter to a more sensible place

<details><summary>Searchable metadata</summary>

```
trac-id__1489 1489
type__enhancement enhancement
reporter__pupeno pupeno
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1140051898000000 1140051898000000
changetime__1214665306000000 1214665306000000
version__ 
owner__ 
cc__pupeno
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="pupeno's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pupeno commented|
|-|-|

```
#!html
<pre>
On "Designing Twisted            
Applications" (http://twistedmatrix.com/projects/core/documentation/howto/design.html)            
for me, it would seem more natural to put the interface, IQuoter,  next to the          
protocol and factory implementation, QOTD(Protocol) and QOTDFactory(Factory),      
instead of the concrete implementation of a Quoter, StaticQuoter and 
FortuneQuoter. 
That is because there will be one protocol implementation, which is a reusable 
library while there would be many implementations of programs using that 
library and all of them would need to implement IQuoter. 
Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to z3p|
|-|-|

Yes, that seems sensible.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

I moved the interface and made a few other very small fixes to things nearby.  `TwistedQuotes` needs some serious attention and more work than I'm going to give it now, though.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix set owner to @exarkun|
|-|-|

Looks good. I even ran the code to test it.


+1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [24129]) Apply twistedquotes.diff, improving the organization of code in the quote example

Author: exarkun
Reviewer: radix
Fixes: [#11032](https://github.com/twisted/twisted/issues/11032)

Move the `IQuoter` interface to be close to the factory which uses it,
instead of being close to the one implementation of it.  This more
closely resembles the organization of a real program, where the user
of an interface does not know or care where the implementation comes
from, only that it behaves correctly, and where there may be several
implementations of the interface in different places.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
